Create reusable `CopyButton` component

rajku-dev opened the pull request:
## Proposed Changes

- Fixes #9494
- edit `copyToClipboard` util to handle null copy operations
- Edit `AssetWarrantyCard` , `SBOMViewer` , `ShiftDetails` to use `copyToClipboard` utility

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new `CopyButton` component for improved clipboard functionality across various components.
	- Enhanced user feedback with tooltips indicating copy status.
	- Added localization support for the "Copy to clipboard" message.

- **Bug Fixes**
	- Streamlined the copying mechanism by removing unnecessary state management and timeout handling, improving user experience.
	- Simplified the structure of components by replacing manual copy implementations with the new `CopyButton`.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rajku-dev committed:
refactor all copy functionalities

rithviknishad commented on src/Utils/utils.ts:
```diff
@@ -564,6 +564,10 @@ export function omitBy<T extends Record<string, unknown>>(
 }
 
 export const copyToClipboard = async (content: string) => {
+  if (!content) {
+    Notification.Error({ msg: "Nothing to copy here!" });
+    return;
+  }
```
like said in previous PR, let's not even give the user the option to copy, when we already know there's nothing to copy.

Why make the user click copy and tell them there's nothing to copy, when we already know there's nothing to copy?

```suggestion
```

rajku-dev replied:
i added this due to serial number being defaults to "--" , I'll make the changes as suggested
if user tries to copy that ("--"), nothing would be copied to clipboard

rithviknishad replied:
disable the button would be better

rajku-dev replied:
made the changes as requested.

rithviknishad commented on src/components/Assets/AssetWarrantyCard.tsx:
```diff
@@ -45,19 +43,18 @@ export default function AssetWarrantyCard(props: { asset: AssetData }) {
-                        <CareIcon icon="l-copy" className="text-lg" />
-                      )}
-                    </CopyToClipboard>
+                  <button
+                    className="tooltip tooltip-bottom"
+                    onClick={() => {
+                      copyToClipboard(details[key as keyof typeof details]);
+                      setIsCopied(true);
+                      setTimeout(() => setIsCopied(false), 2500);
+                    }}
+                  >
+                    <CareIcon
+                      icon={isCopied ? "l-check" : "l-copy"}
+                      className="text-lg"
+                    />
```
consider making this a reusable component, since we are duplicating the logic for isCopied with timeouts now.

rajku-dev replied:
can we use shadcn tooltip inside that component as we are having tooltips in pretty much every copy button @rithviknishad

rithviknishad replied:
yup, go ahead!

rajku-dev committed:
Update src/Utils/utils.ts

Co-authored-by: Rithvik Nishad <rithvikn2001@gmail.com>

rajku-dev committed:
create CopyButton component

rithviknishad commented on src/components/Common/CopyButton.tsx:
```diff
@@ -0,0 +1,58 @@
+
+import { copyToClipboard } from "@/Utils/utils";
+
+interface CopyButtonProps {
+  content: string;
+  tooltipContent?: string;
+  btnContent?: string;
+  resetDuration?: number;
+  iconClassName?: string;
+  btnClassName?: string;
+  icons?: {
+    copied: IconName;
+    copy: IconName;
+  };
+}
```
Let's keep it super simple.

```suggestion
interface CopyButtonProps {
  content: string;
}
```

rajku-dev replied:
even i started with that but..
we might still need btnContent and btnClassName, for SBOM viewer

also some copy icons are bigger than others , so we might need iconClassName for that

are you sure we don't need custom tooltip content? @rithviknishad

rithviknishad replied:
let's use CVA for the size variant for achieving sizes. that should eliminate the need for btnClassName and iconClassName.

btnContent can be renamed to `children` instead.

rajku-dev committed:
use CopyButton component across

rithviknishad commented on src/components/ui/tooltip.tsx:
```diff
@@ -26,7 +26,7 @@ const TooltipComponent = React.forwardRef<
           sideOffset={sideOffset}
           className={cn(
-            "z-50 overflow-hidden rounded-md bg-gray-900 px-3 py-1.5 text-xs text-gray-50 animate-in fade-in-0 zoom-in-95 data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=closed]:zoom-out-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 dark:bg-gray-50 dark:text- ...
+            "z-50 overflow-hidden rounded-md bg-gray-800 px-3 py-1.5 text-sm text-gray-50 animate-in fade-in-0 zoom-in-95 data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=closed]:zoom-out-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 dark:bg-gray-50 dark:text- ...
```
let's not change this

rajku-dev replied:
made the changes as requested.

rajku-dev committed:
disable serial copyBtn

rajku-dev committed:
revert tooltip class

rajku-dev committed:
use cva instead

rajku-dev committed:
Update src/components/Licenses/SBOMViewer.tsx

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

rajku-dev committed:
remove console logs

rithviknishad commented on src/components/Licenses/SBOMViewer.tsx:
```diff
@@ -182,19 +177,19 @@ const BOMDisplay: React.FC = () => {
-            text={JSON.stringify(bomData, null, 2)}
-            onCopy={handleCopy}
+          <CopyButton
+            content={(() => {
+              try {
+                return JSON.stringify(bomData, null, 2);
+              } catch (error) {
+                return '';
+              }
+            })()}
```
```suggestion
            content={JSON.stringify(bomData, null, 2)}
```

rithviknishad commented on src/components/Shifting/ShiftDetails.tsx:
```diff
@@ -42,20 +41,11 @@ export default function ShiftDetails(props: { id: string }) {
-      </a>
+      <CopyButton
+        content={copyContent(data)}
+        tooltipContent="Copy Details to Clipboard"
```
```suggestion
```

rithviknishad commented on src/components/Assets/AssetWarrantyCard.tsx:
```diff
@@ -44,23 +30,13 @@ export default function AssetWarrantyCard(props: { asset: AssetData }) {
+                  details[key as keyof typeof details] && (
+                    <CopyButton
+                      content={details[key as keyof typeof details] || ""}
+                      tooltipContent="Copy to clipboard"
```
```suggestion
```

rajku-dev committed:
drop error handling

Co-authored-by: Rithvik Nishad <rithvikn2001@gmail.com>

rajku-dev committed:
Update src/components/Shifting/ShiftDetails.tsx

Co-authored-by: Rithvik Nishad <rithvikn2001@gmail.com>

rajku-dev committed:
Update src/components/Assets/AssetWarrantyCard.tsx

Co-authored-by: Rithvik Nishad <rithvikn2001@gmail.com>

rajku-dev commented:
@rithviknishad, can we pass the success/error messages to this utility via this component?
This way, we can handle custom notifications and translations as well.

rithviknishad commented:
i don't understand what error would come up when copying

rajku-dev commented:
> i don't understand what error would come up when copying, the only notification that needs to be there is "Copied to clipboard" right?

that's a point. 
If any further changes need to be made please let me know

rithviknishad commented:
lint checks failing

rajku-dev committed:
fix lint

rajku-dev committed:
Merge branch 'develop' into issue/9494/revamp-clipboard

nihal467 committed:
Merge branch 'develop' into issue/9494/revamp-clipboard

rajku-dev committed:
Merge branch 'develop' into issue/9494/revamp-clipboard

rajku-dev committed:
add CopyButton in DoctorVideoSlideover

rajku-dev committed:
Merge branch 'issue/9494/revamp-clipboard' of https://github.com/rajku-dev/care_fe into issue/9494/revamp-clipboard

rajku-dev commented:
forgot to add that in `DoctorVideoSlideover` , made the changes

rajku-dev commented:
> forgot to add the `CopyButton` in `DoctorVideoSlideover` , made the changes now

rithviknishad commented on src/components/Facility/DoctorVideoSlideover.tsx:
```diff
@@ -295,27 +294,10 @@ function UserListItem({ user }: { user: UserAnnotatedWithGroup }) {
-                </span>
-              </a>
+              <CopyButton
+                content={user.alt_phone_number || ""}
```
can we disable the button if there's no content?

```suggestion
                content={user.alt_phone_number}
```

rajku-dev committed:
add disable logic

rithviknishad commented on src/components/Facility/DoctorVideoSlideover.tsx:
```diff
@@ -295,27 +294,11 @@ function UserListItem({ user }: { user: UserAnnotatedWithGroup }) {
-                </span>
-              </a>
+              <CopyButton
+                disabled={!user.alt_phone_number}
```
let's not add a disable prop here, instead let the component itself just disable if there's no content.

```suggestion
```

rajku-dev committed:
fix disable logic

rajku-dev committed:
fix lint

rajku-dev committed:
format with prettier

rithviknishad commented on src/components/Assets/AssetWarrantyCard.tsx:
```diff
@@ -44,23 +30,12 @@ export default function AssetWarrantyCard(props: { asset: AssetData }) {
+                {key === "Serial Number" &&
+                  details[key as keyof typeof details] && (
+                    <CopyButton
+                      content={details[key as keyof typeof details] || ""}
+                    />
```
adjust content prop to accept undefined too, instead let the component handle the hiding logic

```suggestion
                      content={details[key as keyof typeof details]}
                    />
```

rithviknishad commented on src/components/Facility/DoctorVideoSlideover.tsx:
```diff
@@ -295,27 +298,12 @@ function UserListItem({ user }: { user: UserAnnotatedWithGroup }) {
-                  />
-                </span>
-              </a>
+              {user.alt_phone_number && (
+                <CopyButton
+                  content={user.alt_phone_number}
+                  tooltipContent={t("copy_phone_number")}
+                />
+              )}
```
let's handle this logic inside Copy Button since this behaviour is expected on all usages

rajku-dev committed:
Move content handling logic to CopyButton component

rajku-dev committed:
format with prettier

rajku-dev committed:
format ShiftDetails

rithviknishad commented on src/components/Common/CopyButton.tsx:
```diff
@@ -0,0 +1,54 @@
+
+  return (
+    <TooltipProvider>
+      <TooltipComponent content={isCopied ? "Copied!" : tooltipContent}>
```
do follow the merge checklist in the PR comment.

add to i18n strings.

refer: https://github.com/ohcnetwork/care_fe?tab=readme-ov-file#translations

rajku-dev committed:
add translations

rajku-dev committed:
edit CopyButton

rajku-dev committed:
Merge branch 'develop' into issue/9494/revamp-clipboard

rithviknishad commented on src/components/Common/CopyButton.tsx:
```diff
@@ -0,0 +1,58 @@
+
+const CopyButton = ({
+  content,
+  tooltipContent = "copy_to_clipboard",
```
```suggestion
  tooltipContent,
```

rithviknishad commented on src/components/Common/CopyButton.tsx:
```diff
@@ -0,0 +1,58 @@
+  return (
+    <TooltipProvider>
+      <TooltipComponent
+        content={isCopied ? t("copied_to_clipboard") : t(tooltipContent)}
```
```suggestion
        content={isCopied ? t("copied_to_clipboard") : tooltipContent ?? t("copy_to_clipboard")}
```

rajku-dev committed:
Update src/components/Common/CopyButton.tsx

Co-authored-by: Rithvik Nishad <rithvikn2001@gmail.com>

rajku-dev committed:
Update src/components/Common/CopyButton.tsx

Co-authored-by: Rithvik Nishad <rithvikn2001@gmail.com>

bodhish commented:
@rajku-dev please complete the Merge Checklist

rajku-dev committed:
fix lint

rajku-dev commented:
> @rajku-dev please complete the Merge Checklist

@bodhish Sir, I have formatted the code using Prettier to resolve the linting errors. I will ensure to take care of this in future commits as well.

@rithviknishad, please let me know if any further changes are required.

Thank you!

rajku-dev committed:
Merge branch 'develop' into issue/9494/revamp-clipboard

Jacobjeevan reviewed (changes requested):
Hmm clicking on the clipboard in DoctorVideoSlideOver also opens up (the dialog for/or) Whatsapp. Right now, we just have that trigger if user clicks anywhere in the card. I would suggest removing that.

And Clipboard isn't disabled on null (check AssetWarrantyCard as an example).

rajku-dev committed:
Handle conditional rendering of CopyButton in AssetWaarantyCard  | Fix Double action on clicking CopyButton on Doctor Connect Card

rajku-dev committed:
Merge branch 'issue/9494/revamp-clipboard' of https://github.com/rajku-dev/care_fe into issue/9494/revamp-clipboard

rajku-dev commented:
> Hmm clicking on the clipboard in DoctorVideoSlideOver also opens up (the dialog for/or) Whatsapp. Right now, we just have that trigger if user clicks anywhere in the card. I would suggest removing that.
> 
> And Clipboard isn't disabled on null (check AssetWarrantyCard as an example).

@Jacobjeevan  i have made the changes as requested
- Removed anchor tag wrapping in DoctorConnectCard.
-  Implemented conditional rendering for CopyButton in AssetWarrantyCard to display only when serial number is available

rajku-dev committed:
Merge branch 'develop' into issue/9494/revamp-clipboard

rithviknishad commented on src/components/Common/CopyButton.tsx:
```diff
@@ -0,0 +1,62 @@
+  const [isCopied, setIsCopied] = useState(false);
+  const { t } = useTranslation();
+
+  if (content === undefined) return null;
```
```suggestion
  if (!content) return null;
```

Jacobjeevan reviewed (approved):
Implement the suggestion Rithvik mentioned above, otherwise lgtm.

rajku-dev committed:
Update src/components/Common/CopyButton.tsx

Co-authored-by: Rithvik Nishad <rithvikn2001@gmail.com>

rajku-dev commented:
> Implement the suggestion Rithvik mentioned above, otherwise lgtm.

made the required changes

nihal467 commented:
@rajku-dev fix the cypress test related to your PR

rajku-dev commented:
> @rajku-dev fix the cypress test related to your PR
Hi, @nihal467 
The cypress software test for login is failing, with [these](https://github.com/ohcnetwork/care_fe?tab=readme-ov-file#-staging-api-credentials) credentials, after connecting the frontend with local backend setup(docker) 
REACT_CARE_API_URL=http://127.0.0.1:9000, please help me with this
[docs](https://github.com/ohcnetwork/care#self-hosting)

https://github.com/user-attachments/assets/9da9d6a0-5546-4bcf-8c16-9e342cfae983

rajku-dev committed:
fix cypress test

rajku-dev committed:
fix card alignment

rajku-dev commented:
running individual cypress test (cloud api) in chrome for debugging purposes, it is failing due to a different reason compared to it failing here on github
![image](https://github.com/user-attachments/assets/dbb90d45-5b40-40eb-867e-da0514bd1083)
[not able to find the specified patient name "Dummy Patient Eleven"]

rajku-dev committed:
Merge branch 'develop' into issue/9494/revamp-clipboard

rajku-dev commented:
> @rajku-dev fix the cypress test related to your PR

@nihal
All Cypress tests related to this PR have been fixed on my end.
Thank you
